scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala

```diff
@@ -59,6 +61,25 @@ object BeamOp extends Serializable {
   )(implicit ordK: Ordering[K], kryoCoder: KryoCoder): PCollection[KV[K, java.lang.Iterable[U]]] = {
     reduceFn match {
       case ComposedMapGroup(f, g) => planMapGroup(planMapGroup(pcoll, f), g)
+      case EmptyGuard(MapValueStream(SumAll(pqm: PriorityQueueMonoid[V]))) =>
+        pcoll.apply(MapElements.via(
+          new SimpleFunction[KV[K, java.lang.Iterable[V]], KV[K, java.lang.Iterable[U]]]() {
+            override def apply(input: KV[K, lang.Iterable[V]]): KV[K, java.lang.Iterable[U]] = {
+              // We are not using plus method defined in PriorityQueueMonoid as it is mutating
```

Comment:
I think you also need to match on:

https://github.com/twitter/scalding/blob/e1988110f61a2849a077b92250e1e1d58511cd90/scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala#L271

since I guess this restriction is also enforced on the mapside operations... 

You'll need a slightly different approach there (basically a custom cache).

Reply:
Tried with a unit test and it failed because of mutation. Instead of overwriting the `put` method in the `SummingCache` I was hoping to pattern match on Semigroup and overwrite the plus method.

`  case class ImmutablePQMonoid[T](
    size: Int
  )(implicit ord: Ordering[T]) extends PriorityQueueMonoid[T](size) {
    override def plus(left: PriorityQueue[T], right: PriorityQueue[T]): PriorityQueue[T] = {
      super.build(left.iterator().asScala.toIterable ++ right.iterator().asScala.toIterable)
    }
  }
`
But the same issue is there. I cannot access the size field.

Reply:
Or maybe this?

```
  case class ImmutablePQMonoid[T](
    pqm: PriorityQueueMonoid[T]
  )(implicit ord: Ordering[T]) extends Monoid[PriorityQueue[T]] {
    override def zero: PriorityQueue[T] = pqm.zero

    override def plus(
      l: PriorityQueue[T],
      r: PriorityQueue[T]
    ): PriorityQueue[T] = ???
  }
```

Reply:
why not just pattern match on this monoid and not use it at all? Like, as a short fix, you could just make sumByLocalKeys a no-op for this particular monoid in the short term, but for a better fix you could implement a summingcache for this particular monoid.

I think making a full copy of the priority queue each time will be a perf killer and I doubt it will be better than just not using a mapside cache at all...

The more we talk about this, the more it seems like maybe copying the immutable heap in from cats-collections is the right move (or send a PR to algebird and we can publish a new version and use it from there, either way...)

Seems like we keep getting paper cuts here.